piff/optical_model.py

```diff
@@ -61,92 +86,123 @@ class Optical(Model):
+    :param sigma            Gaussian CCD diffusion in arcsec [default:0.]
 
     Since there are a lot of parameters here, we provide the option of setting many of them
     from a template value.  e.g. template = 'des' will use the values stored in the dict
-    piff.optical_model.optical_templates['des'].
+    piff.optatmo_model.aperture_templates['des'].
 
-    :param template:        A key word in the dict piff.optical_model.optical_template to use
-                            for setting values of these parameters.  [default: None]
+    :param template :       A key word in the dict piff.optatmo_model.aperture_template to use
+                            for setting values of these aperture parameters.  [default: None]
+    :param gsparams:        A key word in the dict piff.optatmo_model.gsparams_template to use
+                            for setting values of these Galsim parameters.  [default: None]
 
     If you use a template as well as other specific parameters, the specific parameters will
     override the values from the template.  e.g.  to simulate what DES would be like at
     lambda=1000 nm (the default is 700), you could do:
 
-            >>> model = piff.OpticalModel(template='des', lam=1000)
+            >>> model = piff.Optical(template='des', lam=1000)
+
+    Note that the Zernike coeffients (zernike_coeff), Atmospheric parameters (ie. r0,L0) and Shear (g1,g2) are
+    fitted parameters, passed via the arguments to getProfile.
     """
-    _method = 'no_pixel'
+    _method = 'auto'
+    _model_can_be_offset = True
     _centered = True
-    _model_can_be_offset = False
 
-    def __init__(self, template=None, logger=None, **kwargs):
-        logger = galsim.config.LoggerWrapper(logger)
+    def __init__(self, template=None, gsparams=None, atmo_type='VonKarman', logger=None, **kwargs):
+        self.logger = galsim.config.LoggerWrapper(logger)
+        
         # If pupil_angle and strut angle are provided as strings, eval them.
         for key in ['pupil_angle', 'strut_angle']:
             if key in kwargs and isinstance(kwargs[key],str):
                 kwargs[key] = eval(kwargs[key])
 
-        # Copy over anything from the template dict, but let the direct kwargs override anything
-        # in the template.
+        # Copy over anything from the aperture and gsparams template dict, but let the direct kwargs override anything
         self.kwargs = {}
         if template is not None:
             if template not in optical_templates:
-                raise ValueError("Unknown template specified: %s"%template)
+                raise ValueError("Unknown aperture template specified: %s"%template)
             self.kwargs.update(optical_templates[template])
-        # Do this second, so specified kwargs override anything from the template
+
+        if gsparams is not None:
+            if gsparams not in gsparams_templates:
+                raise ValueError("Unknown gsparams template specified: %s"%gsparams)
```

Comment:
The pattern for testing this is to use `assert_raises` to check that the error is raised when you expect it to be raised.  Either as a function or a context.  Examples:
https://github.com/rmjarvis/Piff/blob/releases/1.2/tests/test_pixel.py#L334
https://github.com/rmjarvis/Piff/blob/releases/1.2/tests/test_pixel.py#L1547

Reply:
Added